psyneulink/library/compositions/pytorchcomponents.py

```diff
@@ -0,0 +1,209 @@
+from psyneulink.core.components.functions.transferfunctions import Linear, Logistic, ReLU
+from psyneulink.library.compositions.pytorchllvmhelper import *
+from psyneulink.core.globals.log import LogCondition
+from psyneulink.core import llvm as pnlvm
+
+import torch
+
+__all__ = ['PytorchMechanismWrapper', 'PytorchProjectionWrapper']
+
+def pytorch_function_creator(function, device, context=None):
+    """
+    Converts a PsyNeuLink function into an equivalent PyTorch lambda function.
+    NOTE: This is needed due to PyTorch limitations (see: https://github.com/PrincetonUniversity/PsyNeuLink/pull/1657#discussion_r437489990)
+    """
+    def get_fct_param_value(param_name):
+        val = function._get_current_function_param(
+            param_name, context=context)
+        if val is None:
+            val = getattr(function.defaults, param_name)
+
+        return float(val)
+
+    if isinstance(function, Linear):
+        slope = get_fct_param_value('slope')
+        intercept = get_fct_param_value('intercept')
+        return lambda x: x * slope + intercept
+
+    elif isinstance(function, Logistic):
+        gain = get_fct_param_value('gain')
+        bias = get_fct_param_value('bias')
+        offset = get_fct_param_value('offset')
+        return lambda x: 1 / (1 + torch.exp(-gain * (x + bias) + offset))
+
+    elif isinstance(function, ReLU):
+        gain = get_fct_param_value('gain')
+        bias = get_fct_param_value('bias')
+        leak = get_fct_param_value('leak')
+        return lambda x: (torch.max(input=(x - bias), other=torch.tensor([0], device=device).double()) * gain +
+                            torch.min(input=(x - bias), other=torch.tensor([0], device=device).double()) * leak)
+
+    else:
+        raise Exception(f"Function {function} is not currently supported in AutodiffCompositions!")
+
+class PytorchMechanismWrapper():
+    """
+    An interpretation of a mechanism as an equivalent pytorch object
+    """
+    def __init__(self, mechanism, component_idx, device, context=None):
+        self._mechanism = mechanism
+        self._idx = component_idx
+        self._context = context
+
+        self.function = pytorch_function_creator(mechanism.function, device, context)
+        self._context = context
+        self.value = None
+        self._llvm_value = None
+        self.afferents = []
+        self.efferents = []
+
+        self._target_mechanism = None
+
+    def add_efferent(self, efferent):
+        if not efferent in self.efferents:
+            self.efferents.append(efferent)
```

Comment:
the interface is weird. is it legal to insert the same efferent twice?

Reply:
Shouldn't be legal (would be akin to adding the same projection twice)

Reply:
`set` doesn't have guaranteed iteration order so that might interfere with indices in compilation.
maybe an assert check would suffice.